### ISSUE TITLE
Server side Inactivity logout

### ISSUE THREAD

**Issue by woogerboy21** [Cockatrice/Cockatrice#2208]
It would nice to see a feature (server side) that the client be logged out of a server after a set period of inactivity.  My initial thoughts were a client side set of code that does a simple counter that counts down and is reset in the event of an action is taken.  If the timer ever reaches 0, the client logs out.  Initial client side setting for timer is something high, maybe 2 hours.  

Anyone else have idea's or thoughts on something like this?

Edit:
First few of actions I can think of that would reset the timer would be:
- Send Chat
- Join Room
- Join Game
- Game Action (shuffle, draw etc)

**Comment by Daenyth** [comment]
I hadn't thought of a client side feature before. That would be a good way
to get it started without being locked into a choice of implementation

**Comment by woogerboy21** [comment]
Hmm... so i started to work on something like this.  I have the timer up and going and a setting in settingscache to define the idle time out value.  I also even have a notification when the idle time out happens.  How ever , now that I go to implement the time reset on action functionality I come to a stumping point.  The timer itself is defined in the remote client portion of the client side code similar to where the ping / keepalive lives (as that is were it seems to make since) but in order for me to actually implement something like a signal/slot I have to create an object to the class's that will be wanting to make a timer reset request.  But that really isnt something that should live in the remote client .....

**Comment by woogerboy21** [comment]
Scratch above, think I have a solution that works pretty well :)

**Comment by tritoch** [comment]
I would like the server to be able to adjust this setting.  Not every server needs to kick idlers.

**Comment by Daenyth** [comment]
@tritoch it's planned - not sure if there's a specific ticket. When it gets added to the server we'll probably remove the client side code that does it.

**Comment by Daenyth** [comment]
This can just be the server side ticket

**Comment by woogerboy21** [comment]
So after the various discussions we have had over the last couple of days I wanted to brainstorm this idea.  At first my thoughts were that doing this would put a major load on the server trying to maintain some type of counter for inactivity for each user that is connected.  But thinking a bit more I think this could be accomplished with a single timer thread.  The idea I had was to create a timer that goes off at an interval that is pulled from the servers settings.ini file.  And for any client that hasn't notified the server of an action gets its session closed.  With this thought the following would have to happen:

* Server side code update requiring a timer
* Client/Server communication update of a valid action (whatever this might be, prob like game action or room message)
* On timer timout signal compare current user list with list collected of users performing actions
* Send session event of client disconnect (new session protocol item?)

Whats the thoughts on this idea?  Any one have a better idea?

**Comment by woogerboy21** [comment]
So just waiting for #2239 to commit before putting the PR in for this but if anyone is interested in seeing the branch it can be found at: https://github.com/woogerboy21/Cockatrice/tree/serverside_idle_timeout

^ also would close #402

**Comment by Daenyth** [comment]
I think that the only hard requirement here is what we put into the protocol. So with that in mind I'm open to ideas.

My main other concern is UX-wise, to not surprise users too badly. I think that of people who will get booted by this, drafters are most at risk.

What about including a system to send generic warnings? A dialog could come up similar to the ban dialog. The server could automatically pop a warning window for inactivity. Moderators could manually send warnings to avoid doing the 1-minute-ban trick.

**Comment by skwerlman** [comment]
The generic system for messages should also be able to send server-wide alerts, for things like maintenance notifications.

**Comment by woogerboy21** [comment]
@Daenyth Currently the only real protocol change that is in would be an event that gets sent.  I don't even really have to add the event since the event handler supports a generic catch all that ultimately results in the same looking dialogue with or without the new event type.  But if your wanting something like a dialogue to open on the client that tells them they will be logged out in X amount of time then yes we would have to extend the protocol.   Is that what you would like to see?

@skwerlman A server wide alert?  What did you have in mind with that and what would be the purpose?  The maintenance task would already be in the logs.  Are you suggesting to spam all the current logged in users with a PM about maintenance that would be happening?

**Comment by Daenyth** [comment]
I'd like a generic way for the server or admins to send targeted messages to a user for example:

```
[Server Notice!] <- title

You have been inactive for $time; you will be logged out in $remaining

[OK]
```

The full message text there `You have been....` should be fully customizable, set by the server. I'm willing to accept not being able to translate it with the flexibility it gives us.

**Comment by woogerboy21** [comment]
Does the "OK" let the server know something or just OK and your SOL?

**Comment by skwerlman** [comment]
@woogerboy21 It'd be nice to be able to send an alert ~5 min before a reboot so people who don't read chat won't get surprised when the server suddenly boots them, though maybe a modal isn't the best way to do it.

**Comment by Daenyth** [comment]
Just an "OK" dialog button. The user would then presumably chat or
something to stop being inactive

**Comment by woogerboy21** [comment]
Alright, I think I understand what you had in mind.  Let me see what I can do.

**Comment by woogerboy21** [comment]
@ctrlaltca had a better solution for performing an idle client timeout so I am writing up a new branch.  Thinking through things I like his method better.

**Comment by woogerboy21** [comment]
Alright, I created the newer branch and submitted PR #2259 . It doesnt implement a new timer for maintenance and simply uses the existing ping timer.  So it wont fix #402.  But this method of doing the idle client time out is much simpler and less resource intensive.

**Comment by woogerboy21** [comment]
> What about including a system to send generic warnings? A dialog could come up similar to the ban dialog. The server could automatically pop a warning window for inactivity. Moderators could manually send warnings to avoid doing the 1-minute-ban trick.

We actually already have a protocol item for this.  Its called "NotifyUser".  But se #2259 for details about its behavior.
